### PR TITLE
SBERDOMA-821 Upgrade structure of BillingReceipt

### PR DESCRIPTION
**Pull request by toplenboren**
* Add billing organization as an inlined list
* Refactor BillingReceipt in a way described in Confluence 
* Add validation for toPayDetails and Services

--

There were no format at all.

Now we have discussed future BillingReceipt format with @pahaz and Alexandr Kostylev and this is the implementation. The format is below [RUSSIAN]

```
// Документация по форматам данных биллингового обмена
//
// Звездочкой * помечены обязательные поля. Все остальные поля считаем необязательными
//
// Дополнительно:
// Мультипринципальные квитанции (когда в одном документе несколько получателей) делим на две. 
// ImportID меняем по правилу: ImportID = <ImportId> + __ + <Index>: 241231 -> (241231__1, 241231__2)

// Квитанция
const BillingReceipt = {
   
    id: "str",                  // *id в CRM
    context: {},                // *BillingIntegrationOrganizationContext в CRM
    receipient: {},             // *BillingOrganization в CRM
    property: {},               // *BillingProperty в CRM
    account: {},                // *BillingAccount в CRM

    importId: "str",            // *id платежного документа в биллинге клиента
    period: new Date(),         // *Период. Формируется как date: 1_<месяц>_<год>
    printableNumber: "str",     //  Номер для печати на квитанции

    toPay: "2000",              // *Детализация уровня 1 - только сумма к оплате
    toPayDetails: {             //  Детализация уровня 2 - детализация суммы к оплате
        "formula": "balance + charge + recalculation + privilege + penalty", //  Формула – задает порядок расчета toPay. 
        "charge": "1000",                  //  Начислено (объем * на тариф)
        "measure": null,                   //  Единица измерения
        "volume": null,                    //  Объем потребления
        "tariff": null,                    //  Тариф
        "balance": "-1000",                //  Входящее сальдо / баланс
        "recalculation": "-1000",          //  Перерасчет (кейс: сдал показания, все пересчитали и накопился долг или переплата за прошлые периоды)
        "privilege": "1000",               //  Льготы
        "penalty": "0",                    //  Пени
    },
    
    "services": [             //  Детализация уровня 2.1 – детализация по услугам (только суммы к оплате)
    {
        id: "number",                 //  Нужен для аналитики
        name: "str",                  //  Название -- ГВС, ХВС, Капремонт, Страховка
        toPay: "1000",      
        toPayDetail: {       //  Детализация уровня 3 – Детализация по услугам (с детализацией суммы)
            "formula": "balance + tariff * volume + recalculation + privilege + penalty", //  Формула – задает порядок расчета toPay. 
            "charge": "1000",         //  Начислено (объем * на тариф)
            "measure": null,          //  Единица измерения
            "volume": 10,             //  Объем потребления
            "tariff": 100,            //  Тариф
            "balance": "-1000",       //  Входящее сальдо / баланс
            "recalculation": "-1000", //  Перерасчет (кейс когда сдал показания и все пересчитали и накопился долг за прошлые периоды или переплата)
            "privilege": "1000",      //  Льготы
            "penalty": "0",           //  Пени
        },
    }
    ]
}


// Лицевой счет
const BillingAccount = {
    context: {},         // *BillingIntegrationOrganizationContext в CRM
    property: {},        // *BillingProperty в CRM
    id: "str",           // *id в CRM
    globalId: "str",     // *ЕЛС
    number: "str",       // *Номер аккаунта в биллинге клиента
    unitName: "number",  //  Номер квартиры или комнаты
    meta: {}             //  Специфичные для биллингов и прочие данные
}

// Дом
const BillingProperty = {
    id: "str",           // *id в CRM
    context: {},         // *BillingIntegrationOrganizationContext в CRM
    globalId: "str",     // *Глобальный идентификатор здания. Для РФ - ФИАС
    address: "str",      //  Адрес дома в CRM (Возможно сделаем скрытым полем и всегда будем заполнять из daData?)
    meta: {}             //  Специфичные для биллингов и прочие данные
}

// Организация-получатель платежа (обычно УК или Ресурсник)
const BillingOrganization = {
    id: "str",           // *id в CRM
    TIN: "str",          // *ИНН
    IEC: "str",          // *КПП
    BIC: "str",          // *БИК
    checkNumber: "str"   // *Расчетный счет
}
```

### CODE REVIEW

File: apps/condo/domains/billing/gql.js [hunk 1]
```diff
@@ -35,9 +35,12 @@ const BillingAccountMeter = generateGqlQueries('BillingAccountMeter', BILLING_AC
 const BILLING_ACCOUNT_METER_READING_FIELDS = `{ context ${BILLING_INTEGRATION_ORGANIZATION_CONTEXT_FIELDS} importId property { id } account { id } meter { id } period value1 value2 value3 date raw ${COMMON_FIELDS} }`
 const BillingAccountMeterReading = generateGqlQueries('BillingAccountMeterReading', BILLING_ACCOUNT_METER_READING_FIELDS)
 
-const BILLING_RECEIPT_FIELDS = `{ context ${BILLING_INTEGRATION_ORGANIZATION_CONTEXT_FIELDS} importId property { id } account { id } period raw toPay services meta ${COMMON_FIELDS} }`
+const BILLING_RECEIPT_FIELDS = `{ context ${BILLING_INTEGRATION_ORGANIZATION_CONTEXT_FIELDS} importId property { id } account { id } recipient { id } period raw toPay toPayDetails services ${COMMON_FIELDS} }`
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/billing/gql.js:38]: printableNumber?

Comment (nomerdvadcatpyat) [reply apps/condo/domains/billing/gql.js:38]: and in readme in BillingOrganization missed context

Comment (toplenboren) [reply apps/condo/domains/billing/gql.js:38]: yep, forgot it. Adding....

**Review by nomerdvadcatpyat — CHANGES_REQUESTED** [review]

**Review by nomerdvadcatpyat — APPROVED** [review]

**Review by dkoviazin — APPROVED** [review]
Maybe, you can combine 2 migrations into one

File: apps/condo/domains/billing/schema/BillingOrganization.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        context: INTEGRATION_CONTEXT_FIELD,
+
+        tin: {
+            schemaDoc: 'Taxpayer Identification Number. In Russia: INN',
+            type: Text,
+            isRequired: true,
+        },
+
+        iec: {
+            schemaDoc: ' Industrial Enterprise Classification. In Russia: KPP',
```
Comment (pahaz) [inline apps/condo/domains/billing/schema/BillingOrganization.js:27]:
```suggestion
            schemaDoc: 'Industrial Enterprise Classification. In Russia: KPP',
```

File: apps/condo/domains/billing/schema/BillingOrganization.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+const BillingOrganization = new GQLListSchema('BillingOrganization', {
+    schemaDoc: 'An organization which can accept payments from the BillingAccount',
+    fields: {
+        dv: DV_FIELD,
+        sender: SENDER_FIELD,
+
+        context: INTEGRATION_CONTEXT_FIELD,
+
+        tin: {
```
Comment (pahaz) [inline apps/condo/domains/billing/schema/BillingOrganization.js:20]: it's better to move it inside meta

File: apps/condo/domains/billing/schema/BillingReceipt.js [hunk 4]
```diff
@@ -18,36 +20,43 @@ const BillingReceipt = new GQLListSchema('BillingReceipt', {
...
+            schemaDoc: 'Total sum to pay. Usually counts as the sum of all services. Detail level 1.',
             type: Text,
             isRequired: true,
         },
 
-        services: {
-            schemaDoc: 'Structured items in the receipt obtained from the `billing data source`. Amount of payment is required for use in the `receipt template`. ' +
-                'Structure example: {"Maintenance and repair": {"name": "maintenance-and-repair", "payment": "1129.17", "currency": "RUB", "formula": "payment = tariff * volume + recalculation + privilege + penalty", "tariff": "19.95", "volume": "56.6", "isByMeter": true, "recalculation": "0.0", "privilege": "0.0", "penalty": "0"}, ...}',
+        toPayDetails: {
+            schemaDoc: 'Sum to pay details. Detail level 2',
```
Comment (pahaz) [inline apps/condo/domains/billing/schema/BillingReceipt.js:44]: it's better to add more docs

File: apps/condo/domains/billing/schema/fields.js [hunk 5]
```diff
@@ -51,7 +61,7 @@ const BILLING_ACCOUNT_METER_FIELD = {
 }
 
 const PERIOD_FIELD = {
-    schemaDoc: 'Period date (01.2020, 02.2020, ...)',
+    schemaDoc: 'Period date: Generated on template 01_<month>_<year>',
     type: CalendarDay,
     isRequired: true,
     // TODO(pahaz): validate it
```
Comment (pahaz) [inline apps/condo/domains/billing/schema/fields.js:67]: it's better to add validation

File: apps/condo/domains/billing/schema/BillingOrganization.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+ */
+
+const { Text } = require('@keystonejs/fields')
+const { GQLListSchema } = require('@core/keystone/schema')
+const { historical, versioned, uuided, tracked, softDeleted } = require('@core/keystone/plugins')
+const { SENDER_FIELD, DV_FIELD } = require('@condo/domains/common/schema/fields')
+const access = require('@condo/domains/billing/access/BillingOrganization')
+const { INTEGRATION_CONTEXT_FIELD } = require('./fields')
+
+const BillingOrganization = new GQLListSchema('BillingOrganization', {
```
Comment (pahaz) [inline apps/condo/domains/billing/schema/BillingOrganization.js:12]: Billing Payment Recipient

**Pull request merged**